torch/distributed/_shard/checkpoint/state_dict_saver.py

```diff
@@ -123,32 +129,75 @@ def save_state_dict(
         >>>     storage_writer=fs_stroage_writer,
         >>> )
     """
-    (
-        metadata,
-        bytes_write_requests,
-        tensor_write_requests,
-    ) = _prepare(state_dict)
-
-    is_rank0 = not dist.is_initialized() or dist.get_rank(process_group) == 0
-    if is_rank0:
-        storage_writer.prepare()
+    is_coordinator = not dist.is_initialized() or dist.get_rank(process_group) == coordinator_rank
+    exceptions: List[Optional[BaseException]] = [None]
+    if is_coordinator:
+        try:
+            storage_writer.prepare()
+        except BaseException as e:
+            exceptions = [e]
 
     # Writing can only start once prepare has finished
     if dist.is_initialized():
-        dist.barrier(process_group)
-
-    combined_writes: List[Union[TensorWriteRequest, BytesWriteRequest]] = []
-    combined_writes.extend(tensor_write_requests)
-    combined_writes.extend(bytes_write_requests)
+        dist.broadcast_object_list(exceptions, group=process_group, src=coordinator_rank)
+
+    if exceptions[0] is not None:
+        raise CheckpointException("failed to prepare storage", {coordinator_rank : exceptions[0]})
+
+    rank_write_error: Optional[BaseException]
+    try:
+        (
+            metadata,
+            bytes_write_requests,
+            tensor_write_requests,
+        ) = _prepare(state_dict, process_group)
+
+        combined_writes: List[Union[TensorWriteRequest, BytesWriteRequest]] = []
+        combined_writes.extend(tensor_write_requests)
+        combined_writes.extend(bytes_write_requests)
+
+        storage_writer.prepare_storage(combined_writes)
+        bytes_futures = storage_writer.write_bytes(bytes_write_requests)
+        tensor_futures = storage_writer.write_tensors(tensor_write_requests)
+        torch.futures.wait_all([bytes_futures, tensor_futures])
+        rank_write_error = None
+    except BaseException as e:
+        rank_write_error = e
+
+    all_errors: List[Optional[BaseException]]
+    # collect all write errors
+    if dist.is_initialized():
```

Comment:
All of this code becomes a lot simpler if we just assert that `dist.is_initialized()`.